readme.md
# FooBar

Example of implementation of encrypted package for modstore.pro repository.

To clone full package with submodules, please, use this command:
```
git clone https://github.com/Alroniks/modx-foo-bar-extra.git foorbar --recursive
```


docs/changelog.txt
This file shows the changes in recent releases of FooBar.

FooBar 1.0.0-pl (TDB)
===================================
- Nothing


_build/helpers/KeyRequester.php
<?php
/**
 * Copyright (c) 2018 Alroniks Experts LLC
 * @author: Ivan Klimchuk <ivan@klimchuk.com>
 * @package: demoshop
 */

include_once __DIR__ . '/ArrayXMLConverter.php';

/**
 * Class KeyRequester
 * Класс для получения пароля шифрования пакета от магазина дополнения
 * В конкретном частном случае - modstore.pro
 */
class KeyRequester
{
    /** @var string Ключ API, выдаваемый магазином после оплаты. Для автора - генерируется бесплатно. */
    const PARAM_API_KEY = 'api_key';
    /** @var string Имя пользователя для доступа к магазину. В случае modstore.pro - email пользователя. */
    const PARAM_USERNAME = 'username';
    /** @var string Хост сайта. В случае реального MODX - выдается домен, но можно заменить на mysite.dev или mysite.docker.*/
    const PARAM_HTTP_HOST = 'http_host';
    /** @var string Имя пакета (бещ версии) */
    const PARAM_PACKAGE = 'package';
    /** @var string Версия пакета вида 1.0.0-pl */
    const PARAM_VERSION = 'version';
    /** @var string Версия протокола взаимодействия с репозиторием. Обычно 1.1, но это новые специальные запросы, поэтому 2.0.0 */
    const PARAM_VEHICLE_VERSION = 'vehicle_version';

    private $xpdo;

    /** @var array Массив с параметрами для доступа к репозиторию */
    private $params = [];

    /**
     * KeyRequester constructor.
     * @param $xpdo
     * @param array $params
     */
    public function __construct(& $xpdo, $params = [])
    {
        $this->xpdo= & $xpdo;

        $this->params = [
            'api_key' => '',
            'username' => '',
            'http_host' => 'anysite.docker',
            'package' => '',
            'version' => '',
            'vehicle_version' => '2.0.0'
        ];

        // Переопределяем параметры теми, которые пришли из кода, которые использует этот класс.
        $this->params = array_merge($this->params, $params);
    }

    /**
     * @return string
     */
    public function getKey()
    {
        // Отправляем запрос к репозиторию и конвертируем ответный XML в массив для удобной работы
        $answer = ArrayXMLConverter::toArray($this->request());

        // Если есть сообщение, значит это сообщение об ошибке.
        // Выводим сообщение и прекращаем работу скрипта сборки (так как без пароля она бесмысслена)
        if (isset($answer['message'])) {
            $this->xpdo->log(xPDO::LOG_LEVEL_ERROR, $answer['message']);
            echo $answer['message'];
            exit;
        }

        // Если сообщений нет, возвращаем ключ, т.е. пароль, который сгенерировал и выдал репозиторий.
        return $answer['key'];
    }

    /**
     * @return string
     */
    protected function request()
    {
        // Стандартный запрос из кода к удаленному сервису через библиотеку cURL.
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, 'https://modstore.pro/extras/package/encode');
        curl_setopt($ch, CURLOPT_POST, 1);
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/xml']);
        curl_setopt($ch, CURLOPT_POSTFIELDS, ArrayXMLConverter::toXML($this->params, 'request'));
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
        $result = trim(curl_exec($ch));
        curl_close($ch);

        return $result;
    }
}


_build/data/transport.settings.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * System settings for package
 *
 * @author Ivan Klimchuk <ivan@klimchuk.com>
 * @package foobar
 */

// Префикс для всех системных настроек, которые будут установлены в пакете
// т.е. foobar_my_system_setting
$prefix = 'foobar_';

// Список объектов системных настроек
$list = [
    'system_setting' => [
        'xtype' => 'textfield',
        'value' => 'barfoo'
    ]
];

$settings = [];
foreach ($list as $k => $v) {
    // создание объекта с необходимыми полями
    // и формирование массива объектов для добавления в пакет
    $setting = new modSystemSetting($xpdo);
    $setting->fromArray(array_merge([
        'key' => $prefix . '_' . $k,
        'namespace' => $prefix,
        'editedon' => null
    ], $v), '', true, true);

    $settings[] = $setting;
}

return $settings;


_build/helpers/ArrayXMLConverter.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

// Класс полностью скопирован с самого MODX для унификации обработки XML.

/**
 * Utility class for array-to-XML transformations.
 * @note Moved from original MODX sources for get independence from installed MODX
 * @package modx
 * @subpackage rest
 */
class ArrayXMLConverter
{
    /**
     * The main function for converting to an XML document.
     * Pass in a multi dimensional array and this recursively loops through and builds up an XML document.
     * @param array $data
     * @param string $rootNodeName - what you want the root node to be - defaults to data.
     * @param SimpleXMLElement $xml - should only be used recursively
     * @return string XML
     */
    public static function toXML($data, $rootNodeName = 'ResultSet', &$xml = null)
    {
        // turn off compatibility mode as simple xml throws a wobbly if you don't.
        if (ini_get('zend.ze1_compatibility_mode') == 1) {
            ini_set('zend.ze1_compatibility_mode', 0);
        }

        if (is_null($xml)) {
            $xml = simplexml_load_string('<?xml version="1.0" encoding="UTF-8" standalone="yes"?><' . $rootNodeName . '></' . $rootNodeName . '>');
        }

        // loop through the data passed in.
        foreach ($data as $key => $value) {

            // no numeric keys in our xml please!
            if (is_numeric($key)) {
                $numeric = 1;
                $key = $rootNodeName;
            }

            // delete any char not allowed in XML element names
            $key = preg_replace('/[^a-z0-9\-\_\.\:]/i', '', $key);

            // if there is another array found recursively call this function
            if (is_array($value)) {
                $node = static::isAssoc($value) || $numeric ? $xml->addChild($key) : $xml;

                // recrusive call.
                if ($numeric) {
                    $key = 'anon';
                }
                static::toXml($value, $key, $node);
            } else {

                // add single node.
                $value = htmlentities($value);
                $xml->addChild($key, $value);
            }
        }

        // pass back as XML
        //return $xml->asXML();

        // if you want the XML to be formatted, use the below instead to return the XML
        $doc = new DOMDocument('1.0');
        $doc->preserveWhiteSpace = false;
        $doc->loadXML($xml->asXML());
        $doc->formatOutput = true;

        return $doc->saveXML();
    }

    /**
     * Convert an XML document to a multi dimensional array
     * Pass in an XML document (or SimpleXMLElement object) and this recrusively loops through and builds a representative array
     * @param string $xml - XML document - can optionally be a SimpleXMLElement object
     * @return array ARRAY
     */
    public static function toArray($xml)
    {
        if (is_string($xml)) {
            $xml = new SimpleXMLElement($xml);
        }
        $children = $xml->children();
        if (!$children) {
            return (string) $xml;
        }
        $arr = array();
        foreach ($children as $key => $node) {

            $node = static::toArray($node);

            // support for 'anon' non-associative arrays
            if ($key == 'anon') {
                $key = count($arr);
            }

            // if the node is already set, put it into an array
            if (isset($arr[$key])) {
                if (!is_array($arr[$key]) || $arr[$key][0] == null) {
                    $arr[$key] = array($arr[$key]);
                }
                $arr[$key][] = $node;
            } else {
                $arr[$key] = $node;
            }
        }

        return $arr;
    }

    /**
     * Determine if a variable is an associative array
     * @static
     * @param mixed $array The variable to check
     * @return boolean True if is an array
     */
    public static function isAssoc($array)
    {
        return (is_array($array) && 0 !== count(array_diff_key($array, array_keys(array_keys($array)))));
    }
}


_build/resolvers/resolve.settings.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * System settings resolver
 *
 * @author Ivan Klimchuk <ivan@klimchuk.com>
 * @package foorbar
 */

if (!$object->xpdo && !$object->xpdo instanceof modX) {
    return true;
}

switch ($options[xPDOTransport::PACKAGE_ACTION]) {
    case xPDOTransport::ACTION_UNINSTALL:
        $object->xpdo->removeCollection('modSystemSetting', ['key:LIKE' => 'foobar\_%']);
        break;
}

return true;


_packages/foobar-1.0.0-pl/manifest.php
<?php return array (
  'manifest-version' => '1.1',
  'manifest-attributes' => 
  array (
    'changelog' => 'This file shows the changes in recent releases of FooBar.

FooBar 1.0.0-pl (TDB)
===================================
- Nothing
',
    'license' => 'The MIT License (MIT)
Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
TEXT OF LICENSE
',
    'readme' => '# FooBar
Example of implementation of encrypted package for modstore.pro repository.
',
    'requires' => 
    array (
      'php' => '>=5.5',
      'modx' => '>=2.6',
    ),
  ),
  'manifest-vehicles' => 
  array (
    0 => 
    array (
      'vehicle_package' => 'transport',
      'vehicle_class' => 'xPDOFileVehicle',
      'class' => 'xPDOFileVehicle',
      'guid' => '260ffcbf5a84da7d1fe49bfa1f5a7525',
      'native_key' => NULL,
      'filename' => 'xPDOFileVehicle/8b05e897129aff5d3d651dd318971192.vehicle',
    ),
    1 => 
    array (
      'vehicle_package' => 'transport',
      'vehicle_class' => 'xPDOScriptVehicle',
      'class' => 'xPDOScriptVehicle',
      'guid' => '9a8b493ad3b8f85fe015cb0e12b50451',
      'native_key' => NULL,
      'filename' => 'xPDOScriptVehicle/1b98e1cbfab540910f1c3ce593e458ba.vehicle',
    ),
    2 => 
    array (
      'vehicle_package' => 'transport',
      'vehicle_class' => 'EncryptedVehicle',
      'class' => 'modNamespace',
      'guid' => '53d04102dad11ea22ed1a54c1644465c',
      'native_key' => NULL,
      'filename' => 'modNamespace/864c32b50e215d5669b1e2c1a161cdaa.vehicle',
      'namespace' => 'foobar',
    ),
    3 => 
    array (
      'vehicle_package' => 'transport',
      'vehicle_class' => 'EncryptedVehicle',
      'class' => 'modSystemSetting',
      'guid' => 'ede0625a4561c954699fbd757e3e1fac',
      'native_key' => NULL,
      'filename' => 'modSystemSetting/ec7b3a48e1cd588b60ceeb361c3b30e6.vehicle',
      'namespace' => 'foobar',
    ),
  ),
);

_build/resolvers/resolve.encryption.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

// Данный файл выполняется как PHP скрипт в момент установки пакета
// и регистрирует специальный класс EncryptedVehicle
// который определяет тип транспортного объекта пакета

/** @var xPDOTransport $transport */
$transport->xpdo->loadClass('transport.xPDOObjectVehicle', XPDO_CORE_PATH, true, true);
$transport->xpdo->loadClass('EncryptedVehicle', MODX_CORE_PATH . 'components/' . strtolower($transport->name) . '/', true, true);


_packages/foobar-1.0.0-pl/xPDOFileVehicle/8b05e897129aff5d3d651dd318971192/encryptedvehicle.class.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * Class EncryptedVehicle
 * @author Vasiliy Naumkin
 */
class EncryptedVehicle extends xPDOObjectVehicle
{
    const VERSION = '2.0.0';
    const CIPHER = 'AES-256-CBC';

    public $class = self::class;

    /**
     * @param $transport xPDOTransport
     * @param $object
     * @param array $attributes
     */
    public function put(&$transport, &$object, $attributes = array())
    {
        parent::put($transport, $object, $attributes);

        if (defined('PKG_ENCODE_KEY')) {

            $this->payload['object_encrypted'] = $this->encode($this->payload['object'], PKG_ENCODE_KEY);
            unset($this->payload['object']);

            if (isset($this->payload['related_objects'])) {
                $this->payload['related_objects_encrypted'] = $this->encode($this->payload['related_objects'], PKG_ENCODE_KEY);
                unset($this->payload['related_objects']);
            }

            $transport->xpdo->log(xPDO::LOG_LEVEL_INFO, 'Package encrypted!');
        }
    }

    /**
     * @param $transport xPDOTransport
     * @param $options
     *
     * @return bool
     */
    public function install(&$transport, $options)
    {
        if (!$this->decodePayloads($transport, 'install')) {
            $transport->xpdo->log(xPDO::LOG_LEVEL_ERROR, 'Package can not be decrypted!');
            return false;
        } else {
            $transport->xpdo->log(xPDO::LOG_LEVEL_INFO, 'Package decrypted!');
        }

        return parent::install($transport, $options);
    }

    /**
     * @param $transport xPDOTransport
     * @param $options
     *
     * @return bool
     */
    public function uninstall(&$transport, $options)
    {
        if (!$this->decodePayloads($transport, 'uninstall')) {
            $transport->xpdo->log(xPDO::LOG_LEVEL_ERROR, 'Package can not be decrypted!');
            return false;
        } else {
            $transport->xpdo->log(xPDO::LOG_LEVEL_INFO, 'Package decrypted!');
        }

        return parent::uninstall($transport, $options);
    }

    /**
     * @param array $data
     * @param string $key
     *
     * @return string
     */
    protected function encode($data, $key)
    {
        $ivLen = openssl_cipher_iv_length(self::CIPHER);
        $iv = openssl_random_pseudo_bytes($ivLen);
        $cipher_raw = openssl_encrypt(serialize($data), self::CIPHER, $key, OPENSSL_RAW_DATA, $iv);

        return base64_encode($iv . $cipher_raw);
    }

    /**
     * @param string $string
     * @param string $key
     *
     * @return string
     */
    protected function decode($string, $key)
    {
        $ivLen = openssl_cipher_iv_length(self::CIPHER);
        $encoded = base64_decode($string);

        if (ini_get('mbstring.func_overload')) {
            $strLen = mb_strlen($encoded, '8bit');
            $iv = mb_substr($encoded, 0, $ivLen, '8bit');
            $cipher_raw = mb_substr($encoded, $ivLen, $strLen, '8bit');
        } else {
            $iv = substr($encoded, 0, $ivLen);
            $cipher_raw = substr($encoded, $ivLen);
        }

        return unserialize(openssl_decrypt($cipher_raw, self::CIPHER, $key, OPENSSL_RAW_DATA, $iv));
    }

    /**
     * @param $transport xPDOTransport
     * @param string $action
     *
     * @return bool
     */
    protected function decodePayloads(&$transport, $action = 'install')
    {
        if (isset($this->payload['object_encrypted']) || isset($this->payload['related_objects_encrypted'])) {
            if (!$key = $this->getDecodeKey($transport, $action)) {
                return false;
            }
            if (isset($this->payload['object_encrypted'])) {
                $this->payload['object'] = $this->decode($this->payload['object_encrypted'], $key);
                unset($this->payload['object_encrypted']);
            }
            if (isset($this->payload['related_objects_encrypted'])) {
                $this->payload['related_objects'] = $this->decode($this->payload['related_objects_encrypted'], $key);
                unset($this->payload['related_objects_encrypted']);
            }
        }

        return true;
    }

    /**
     * @param $transport xPDOTransport
     * @param $action
     *
     * @return bool|string
     */
    protected function getDecodeKey(&$transport, $action)
    {
        $key = false;
        $endpoint = 'package/decode/' . $action;

        /** @var modTransportPackage $package */
        $package = $transport->xpdo->getObject('transport.modTransportPackage', [
            'signature' => $transport->signature
        ]);

        if ($package instanceof modTransportPackage) {
            /** @var modTransportProvider $provider */
            if ($provider = $package->getOne('Provider')) {

                $provider->xpdo->setOption('contentType', 'default');
                $params = array(
                    'package' => $package->package_name,
                    'version' => $transport->version,
                    'username' => $provider->username,
                    'api_key' => $provider->api_key,
                    'vehicle_version' => self::VERSION,
                );

                $response = $provider->request($endpoint, 'POST', $params);

                if ($response->isError()) {
                    $msg = $response->getError();
                    $transport->xpdo->log(xPDO::LOG_LEVEL_ERROR, $msg);
                } else {
                    $data = $response->toXml();
                    if (!empty($data->key)) {
                        $key = $data->key;
                    } elseif (!empty($data->message)) {
                        $transport->xpdo->log(xPDO::LOG_LEVEL_ERROR, $data->message);
                    }
                }
            }
        }

        return $key;
    }
}


_build/build.transport.php
<?php
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Ivan Klimchuk <ivan@klimchuk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * FooBar package builder
 *
 * @author Ivan Klimchuk <ivan@klimchuk.com>
 * @package foorbar
 */

// Удаляем стандартные ограничения времени выполнения скрипта,
// чтобы все долгие операции сборки гарантированно выполнились.
set_time_limit(0);

// xPDO требует, чтобы была установлена правильная временная зона в настройках PHP
ini_set('date.timezone', 'Europe/Minsk');

// Здесь описывается блок констант с информацией о нашем пакете.
define('PKG_NAME', 'FooBar'); // Название пакета для людей
define('PKG_NAME_LOWER', strtolower(PKG_NAME)); // Название для репозитория и прочих программ
define('PKG_VERSION', '1.0.0'); // Версия пакета
define('PKG_RELEASE', 'pl'); // Релиз пакета (MODX все еще требует этот рудимент, можно всегда оставлять pl)

// Здесь мы подключаем классы, которые нам понадобятся при сборке пакета.
require_once __DIR__ . '/xpdo/xpdo/xpdo.class.php';
require_once __DIR__ . '/xpdo/xpdo/transport/xpdotransport.class.php';
require_once __DIR__ . '/xpdo/xpdo/transport/xpdovehicle.class.php';
require_once __DIR__ . '/xpdo/xpdo/transport/xpdofilevehicle.class.php';
require_once __DIR__ . '/xpdo/xpdo/transport/xpdoscriptvehicle.class.php';
require_once __DIR__ . '/xpdo/xpdo/transport/xpdoobjectvehicle.class.php';

// Здесь мы подключаем вспомогательные классы, которые используются во время сборки пакета.
require_once __DIR__ . '/helpers/KeyRequester.php'; // Класс для получения пароля для шифрования пакета.
require_once __DIR__ . '/implants/encryptedvehicle.class.php'; // Класс шифровальщика.
// Можно писать имя класса в виде строки вот так 'EncryptedVehicle',
// но удобнее пользоваться новыми возможностями PHP и писать так 'EncryptedVehicle::class'.
// Для этого класс должен быть подключен явно или через автозагрузчик.

// Создаем реальный объект класса xPDO.
// Параметры указывать обязательно, но настоящая базад данных не требуется.
// Будет работать и так, разве что с предупреждениями. Но это никак не влияет на результат.
$xpdo = xPDO::getInstance('db', [
    xPDO::OPT_CACHE_PATH => __DIR__ . '/../cache/',
    xPDO::OPT_HYDRATE_FIELDS => true,
    xPDO::OPT_HYDRATE_RELATED_OBJECTS => true,
    xPDO::OPT_HYDRATE_ADHOC_FIELDS => true,
    xPDO::OPT_CONNECTIONS => [
        [
            'dsn' => 'mysql:host=localhost;dbname=xpdotest;charset=utf8',
            'username' => 'test',
            'password' => 'test',
            'options' => [xPDO::OPT_CONN_MUTABLE => true],
            'driverOptions' => [],
        ]
    ]
]);

// Устанавливаем логирование ошибок. Можете тут поэкспериментировать.
$xpdo->setLogLevel(xPDO::LOG_LEVEL_FATAL);
$xpdo->setLogTarget();

// xPDO знает только о своих классах, а чтобы создавать сущности MODX - нужно покдлючать классы из MODX.
// Но у нас его нет, так что нужно эти классы сделать искуственно. Нам не нужы настоящие классы, только их заглушки,
// чтобы xPDO не ругался на несуществующие классы. Небольшой хак, так как по другому во 2 версии нельзя.
class modNamespace extends xPDOObject {}
class modCategory extends xPDOObject {
    public function getFKDefinition($alias)
    {
        $aggregates = [
            'Plugins' => [
                'class' => 'modPlugin',
                'local' => 'id',
                'foreign' => 'category',
                'cardinality' => 'many',
                'owner' => 'local',
            ]
        ];

        return isset($aggregates[$alias]) ? $aggregates[$alias] : [];
    }
}
class modPlugin extends xPDOObject {
    public function getFKDefinition($alias)
    {
        $aggregates = [
            'PluginEvents' => [
                'class' => 'modPluginEvent',
                'local' => 'id',
                'foreign' => 'pluginid',
                'cardinality' => 'one',
                'owner' => 'local',
            ]
        ];

        return isset($aggregates[$alias]) ? $aggregates[$alias] : [];
    }
}
class modPluginEvent extends xPDOObject {}
class modSystemSetting extends xPDOObject {}
class msPayment extends xPDOObject {}

// Создаем массив с некоторыми настройками для сборки пакета.
// Острой необходимссти нет, но просто удобнее работать, когда пакет большой и нужно что-то поменять.
// Поменять в одном месте намного проще чем в 10 по всему файлу. И ошибок меньше.
$root = dirname(dirname(__FILE__)) . '/';
$sources = [
    'build' => $root . '_build/',
    'data' => $root . '_build/data/',
    'docs' => $root . 'docs/',
    'resolvers' => $root . '_build/resolvers/',
    'validators' => $root . '_build/validators/',
    'implants' => $root . '_build/implants/',
    'plugins' => $root . 'core/components/' . PKG_NAME_LOWER . '/elements/plugins/',
    'assets' => ['components/' . PKG_NAME_LOWER . '/'],
    'core' => ['components/' . PKG_NAME_LOWER . '/'],
];

// Полная сигнатура пакета, т.е. foobar-1.0.0-pl
$signature = join('-', [PKG_NAME_LOWER, PKG_VERSION, PKG_RELEASE]);
// Каталог, куда будет собираться пакет (архив)
$directory = $root . '_packages/';
// Имя файла уже собранного пакета (нужно, чтобы удалять уже собранные перед новой сборкой)
$filename = $directory . $signature . '.transport.zip';

// Если файл пакета уже есть, то удаляем перед новой сборкой
if (file_exists($filename)) {
    unlink($filename);
}
// Кроме файла удаляем и папку с элементами пакета и чистим кеш xPDO (да, кеш - это фишка xPDO, а не MODX)
if (file_exists($directory . $signature) && is_dir($directory . $signature)) {
    $cacheManager = $xpdo->getCacheManager();
    if ($cacheManager) {
        $cacheManager->deleteTree($directory . $signature, true, false, []);
    }
}

$keyRequester = new KeyRequester($xpdo, [
    KeyRequester::PARAM_API_KEY => 'api_key_from_modstore.pro', // ключ из modstore.pro
    KeyRequester::PARAM_USERNAME => 'email@on-mostore.pro', // email аккаунта на modstore.pro
    KeyRequester::PARAM_PACKAGE => PKG_NAME_LOWER, // имя пакета
    KeyRequester::PARAM_VERSION => PKG_VERSION . '-' . PKG_RELEASE // версия пакета
]);

// Когда получили пароль для шифрования, сохраняем его в константу,
// значение которой будет использоваться в классе шифровальщика
// при шифровании уже отдельных элементов пакета.
define('PKG_ENCODE_KEY', $keyRequester->getKey());

// Создаем объект самого транспортного пакета
// Передаем xpdo, сигнатуру и каталог, куда положить файл с пакетом
$package = new xPDOTransport($xpdo, $signature, $directory);

// Сначала добавляем транспортный объект файлового типа
// И в параметрах указываем источник, откуда брать файл с классом (в папке implants нашего пакета)
// а при установке кладем его в папку с нашим компонентов.
// За это отвечает команда, описанная в 'target'
$package->put(new xPDOFileVehicle, [
    'vehicle_class' => 'xPDOFileVehicle',
    'object' => [
        'source' => $sources['implants'] . 'encryptedvehicle.class.php',
        'target' => "return MODX_CORE_PATH . 'components/" . PKG_NAME_LOWER . "/';"
    ]
]);

// Затем, когда файл скопирован, добавляем специальный скрипт,
// который запустится следом за копированием файла и зарегистрирует класс шифровальщика в MODX
$package->put(new xPDOScriptVehicle, [
    'vehicle_class' => 'xPDOScriptVehicle',
    'object' => [
        'source' => $sources['resolvers'] . 'resolve.encryption.php'
    ]
]);

// Теперь начинаем добавлять настоящие элементы в пакет,
// которые после установки будут созданы и доступны в системе

// Создаем объект пространства имен для нашего пакета
$namespace = new modNamespace($xpdo);
$namespace->fromArray([
    'id' => PKG_NAME_LOWER,
    'name' => PKG_NAME_LOWER,
    'path' => '{core_path}components/' . PKG_NAME_LOWER . '/',
]);

// И тут же его добавляем в наш пакет
$package->put($namespace, [
    // в качестве типа транспортного объекта указываем EncryptedVehicle
    // это значит, что данный объект зашифрует свое содержимое
    // и при установке будет пытаться его расшифровать
    'vehicle_class' => EncryptedVehicle::class,
    xPDOTransport::UNIQUE_KEY => 'name',
    xPDOTransport::PRESERVE_KEYS => true,
    xPDOTransport::UPDATE_OBJECT => true,
    xPDOTransport::NATIVE_KEY => PKG_NAME_LOWER,
    'namespace' => PKG_NAME_LOWER
]);

// Подключаем наш массив объектов системных настроек
$settings = include $sources['data'] . 'transport.settings.php';
// И точно так же создаем транспортные объекты для каждой настройки
// и кладем их в пакет с шифрованным типом
foreach ($settings as $setting) {
    $package->put($setting, [
        'vehicle_class' => EncryptedVehicle::class,
        xPDOTransport::UNIQUE_KEY => 'key',
        xPDOTransport::PRESERVE_KEYS => true,
        xPDOTransport::UPDATE_OBJECT => true,
        'class' => 'modSystemSetting',
        'namespace' => PKG_NAME_LOWER
    ]);
}

// Так как пакет у нас простой, у нас тут нет специальных файловых резолверов, поэтому они здесь пропущены
// Но есть вызов резолвера в виде скрипта, который будет выполняться при удалении пакета
// и удалять установленные ранее пакетом системные настройки.
$resolvers = [
    ['type' => 'php', 'source' => $sources['resolvers'] . 'resolve.settings.php']
];

// Здесь в пакет добавляем мета-ифнормацию. Эта информация используется репозиторием и
// механизмом управления пакетами в самом MODX
// Файл изменений для пакета
$package->setAttribute('changelog', file_get_contents($sources['docs'] . 'changelog.txt'));
// Файл с лицензией
$package->setAttribute('license', file_get_contents($sources['docs'] . 'license.txt'));
// Файл описания пакета
$package->setAttribute('readme', file_get_contents($sources['docs'] . 'readme.txt'));
// Если нужно, указываем зависимости для пакета (минимальная версия PHP, MODX и/или наличие других компонентов)
$package->setAttribute('requires', [
    'php' => '>=5.5',
    'modx' => '>=2.6'
]);

// Ну и наконец, упаковываем наш пакет. Результат будет доступен в папке _packages, если вы не меняли настройки выше.
if ($package->pack()) {
    $xpdo->log(xPDO::LOG_LEVEL_INFO, "Package built");
}
